From eeab118e0e913fd0535e7c4334a7b8bcea729a9e Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Mon, 22 Jul 2024 10:02:52 +0200
Subject: [PATCH 1082/1695] drivers: stm32_rifsc: implement errata on RIMUs

When "st,errata-ahbrisab" is set in the device tree, RIMUs cannot hold
the CID0 value on the bus.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git c082a9ec79bd6ce8625f99141b5dea2a35ce7322 drivers: stm32_rifsc: implement errata on RIMUs]
Change-Id: I96cc451605cf8ecc87edd0744fe6fb761e6826c8
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/394636
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Tested-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 core/drivers/firewall/stm32_rifsc.c | 201 +++++++++++++++++++++++++++-
 core/include/drivers/stm32_rifsc.h  |   8 ++
 2 files changed, 208 insertions(+), 1 deletion(-)

diff --git a/core/drivers/firewall/stm32_rifsc.c b/core/drivers/firewall/stm32_rifsc.c
index 964ecea14..0cf470df3 100644
--- a/core/drivers/firewall/stm32_rifsc.c
+++ b/core/drivers/firewall/stm32_rifsc.c
@@ -93,8 +93,10 @@
 					 RIFSC_RISC_SCID_MASK | \
 					 RIFSC_RISC_SEML_MASK)
 
-#define RIFSC_RIMC_MODE_MASK		BIT(2)
+#define RIFSC_RIMC_CIDSEL_MASK		BIT(2)
+#define RIFSC_RIMC_CIDSEL_SHIFT		U(2)
 #define RIFSC_RIMC_MCID_MASK		GENMASK_32(6, 4)
+#define RIFSC_RIMC_MCID_SHIFT		U(4)
 #define RIFSC_RIMC_MSEC_MASK		BIT(8)
 #define RIFSC_RIMC_MPRIV_MASK		BIT(9)
 
@@ -120,8 +122,131 @@ struct rifsc_platdata {
 	struct rimu_cfg *rimu;
 	int nrimu;
 	bool is_tdcid;
+	bool errata_ahbrisab;
 };
 
+struct rimu_risup_pairs {
+	uint32_t rimu_id;
+	uint32_t risup_id;
+};
+
+#if defined(CFG_STM32MP25) || defined(CFG_STM32MP23)
+static const struct rimu_risup_pairs rimu_risup[] = {
+	[0] = {
+		.rimu_id = 0,
+		.risup_id = 0,
+	},
+	[1] = {
+		.rimu_id = 1,
+		.risup_id = STM32MP25_RIFSC_SDMMC1_ID,
+	},
+	[2] = {
+		.rimu_id = 2,
+		.risup_id = STM32MP25_RIFSC_SDMMC2_ID,
+	},
+	[3] = {
+		.rimu_id = 3,
+		.risup_id = STM32MP25_RIFSC_SDMMC3_ID,
+	},
+	[4] = {
+		.rimu_id = 4,
+		.risup_id = STM32MP25_RIFSC_USB3DR_ID,
+	},
+	[5] = {
+		.rimu_id = 5,
+		.risup_id = STM32MP25_RIFSC_USBH_ID,
+	},
+	[6] = {
+		.rimu_id = 6,
+		.risup_id = STM32MP25_RIFSC_ETH1_ID,
+	},
+	[7] = {
+		.rimu_id = 7,
+		.risup_id = STM32MP25_RIFSC_ETH2_ID,
+	},
+	[8] = {
+		.rimu_id = 8,
+		.risup_id = STM32MP25_RIFSC_PCIE_ID,
+	},
+	[9] = {
+		.rimu_id = 9,
+		.risup_id = STM32MP25_RIFSC_GPU_ID,
+	},
+	[10] = {
+		.rimu_id = 10,
+		.risup_id = STM32MP25_RIFSC_DCMIPP_ID,
+	},
+	[11] = {
+		.rimu_id = 11,
+		.risup_id = 0,
+	},
+	[12] = {
+		.rimu_id = 12,
+		.risup_id = 0,
+	},
+	[13] = {
+		.rimu_id = 13,
+		.risup_id = 0,
+	},
+	[14] = {
+		.rimu_id = 14,
+		.risup_id = STM32MP25_RIFSC_VDEC_ID,
+	},
+	[15] = {
+		.rimu_id = 15,
+		.risup_id = STM32MP25_RIFSC_VENC_ID,
+	},
+};
+#endif /* defined(CFG_STM32MP25) || defined(CFG_STM32MP23) */
+#if defined(CFG_STM32MP21)
+static const struct rimu_risup_pairs rimu_risup[] = {
+	[0] = {
+		.rimu_id = 0,
+		.risup_id = 0,
+	},
+	[1] = {
+		.rimu_id = 1,
+		.risup_id = STM32MP21_RIFSC_SDMMC1_ID,
+	},
+	[2] = {
+		.rimu_id = 2,
+		.risup_id = STM32MP21_RIFSC_SDMMC2_ID,
+	},
+	[3] = {
+		.rimu_id = 3,
+		.risup_id = STM32MP21_RIFSC_SDMMC3_ID,
+	},
+	[4] = {
+		.rimu_id = 4,
+		.risup_id = STM32MP21_RIFSC_OTG_HS_ID,
+	},
+	[5] = {
+		.rimu_id = 5,
+		.risup_id = STM32MP21_RIFSC_USBH_ID,
+	},
+	[6] = {
+		.rimu_id = 6,
+		.risup_id = STM32MP21_RIFSC_ETH1_ID,
+	},
+	[7] = {
+		.rimu_id = 7,
+		.risup_id = STM32MP21_RIFSC_ETH2_ID,
+	},
+	[10] = {
+		.rimu_id = 10,
+		.risup_id = STM32MP21_RIFSC_DCMIPP_ID,
+	},
+	[11] = {
+		.rimu_id = 11,
+		.risup_id = 0,
+	},
+	[12] = {
+		.rimu_id = 12,
+		.risup_id = 0,
+	},
+};
+#endif /* defined(CFG_STM32MP21) */
+
 static struct rifsc_driver_data rifsc_drvdata;
 static struct rifsc_platdata rifsc_pdata;
 
@@ -293,6 +418,9 @@ static TEE_Result stm32_rifsc_parse_fdt(const void *fdt, int node,
 			return res;
 	}
 
+	rifsc_pdata.errata_ahbrisab = fdt_getprop(fdt, node,
+						  "st,errata-ahbrisab", NULL);
+
 	return TEE_SUCCESS;
 }
 
@@ -383,6 +511,75 @@ static TEE_Result stm32_risup_setup(struct rifsc_platdata *pdata)
 	return TEE_SUCCESS;
 }
 
+/*
+ * Errata: When CID filtering is enabled on one of RISAB 3/4/5 instances, we
+ * forbid the use of CID0 for any initiator on the bus to handle spurious CID0
+ * transactions on these RAMs.
+ */
+static void stm32_rimu_errata_ahbrisab(struct rifsc_platdata *pdata,
+				       struct rimu_cfg *rimu)
+{
+	unsigned int i = 0;
+
+	if (!pdata->errata_ahbrisab)
+		return;
+
+	for (i = 0; i < ARRAY_SIZE(rimu_risup); i++) {
+		if (rimu->id == rimu_risup[i].rimu_id) {
+			rimu->risup_id = rimu_risup[i].risup_id;
+			break;
+		}
+	}
+
+	if (rimu->attr & RIFSC_RIMC_CIDSEL_MASK) {
+		/* No inheritance mode for this RIMU */
+		if ((rimu->attr & RIFSC_RIMC_MCID_MASK) >>
+		    RIFSC_RIMC_MCID_SHIFT == RIF_CID0) {
+			EMSG("A CID should be set for RIMU %u", rimu->id);
+			if (!IS_ENABLED(CFG_INSECURE))
+				panic();
+			}
+	} else {
+		struct risup_cfg *risup = NULL;
+		uint32_t risup_cidcfgr = 0;
+		int j = 0;
+
+		/* Handle RIMU with no inheritance mode */
+		if (!rimu->risup_id) {
+			EMSG("RIMU%u cannot be set in inheritance mode",
+			     rimu->id);
+			if (!IS_ENABLED(CFG_INSECURE))
+				panic();
+			return;
+		}
+
+		for (j = 0; j < pdata->nrisup; j++) {
+			if (rimu->risup_id == pdata->risup[j].id) {
+				risup = &pdata->risup[j];
+				break;
+			}
+		}
+
+		if (!risup)
+			panic();
+
+		risup_cidcfgr = io_read32(pdata->base +
+					  _RIFSC_RISC_PER0_CIDCFGR +
+					  _OFFSET_PERX_CIDCFGR * risup->id);
+
+		if (!(risup_cidcfgr & RIFSC_RISC_CIDCFGR_CFEN_MASK) ||
+		    (!(risup_cidcfgr & RIFSC_RISC_CIDCFGR_SEM_EN_MASK) &&
+		     ((risup_cidcfgr & RIFSC_RISC_CIDCFGR_SCID_MASK) >>
+		      RIFSC_RISC_CIDCFGR_SCID_SHIFT) == RIF_CID0) ||
+		    (risup_cidcfgr & RIFSC_RISC_CIDCFGR_SEM_EN_MASK &&
+		     risup_cidcfgr & BIT(RIFSC_RISC_CIDCFGR_SEML_SHIFT))) {
+			EMSG("RIMU%u in inheritance mode with CID0", rimu->id);
+			if (!IS_ENABLED(CFG_INSECURE))
+				panic();
+		}
+	}
+}
+
 static TEE_Result stm32_rimu_cfg(struct rifsc_platdata *pdata,
 				 struct rimu_cfg *rimu)
 {
@@ -392,6 +589,8 @@ static TEE_Result stm32_rimu_cfg(struct rifsc_platdata *pdata,
 	if (!rimu || rimu->id >= drv_data->nb_rimu)
 		return TEE_ERROR_BAD_PARAMETERS;
 
+	stm32_rimu_errata_ahbrisab(pdata, rimu);
+
 	if (drv_data->rif_en)
 		io_write32(pdata->base + offset, rimu->attr);
 
diff --git a/core/include/drivers/stm32_rifsc.h b/core/include/drivers/stm32_rifsc.h
index f3d03ad16..18244036a 100644
--- a/core/include/drivers/stm32_rifsc.h
+++ b/core/include/drivers/stm32_rifsc.h
@@ -17,8 +17,16 @@ struct risup_cfg {
 	bool pm_sem;
 };
 
+/**
+ * struct rimu_cfg - RIMU configuration
+ *
+ * @id: ID of the RIMU
+ * @risup_id: ID of the associated RISUP
+ * @attr: RIMU configuration attributes
+ */
 struct rimu_cfg {
 	uint32_t id;
+	uint32_t risup_id;
 	uint32_t attr;
 };
 
-- 
2.47.2

